scipy/spatial/_hausdorff.pyx

```diff
@@ -35,7 +35,11 @@ def directed_hausdorff(const double[:,::1] ar1, const double[:,::1] ar2, seed=0)
     # shuffling the points in each array generally increases the likelihood of
     # an advantageous break in the inner search loop and never decreases the
     # performance of the algorithm
-    rng = np.random.RandomState(seed)
+    if isinstance(seed, int | None):
+        rng = np.random.RandomState(seed)
+    else:
+        # Generator passthrough
+        rng = seed
```

Comment:
There are a few differences between this and the old/usual behavior (e.g. it doesn't look like NumPy integers will be work anymore). To guarantee backward compatibility, I'd suggest branching based on whether `rng` is an instance of `Generator` instead of looking for types that can be passed to `RandomState`.

On the other hand, it would be easiest for the transition if this were to behave like other functions. Can you use [`check_random_state`](https://github.com/scipy/scipy/blob/e24c0cccefb60d927b41ae75a1a772367dd3e340/scipy/_lib/_util.py#L482-L490) or copy that logic here?  It looks like the differences would be:
- `RandomState` (and `Generator`) instances would now be accepted
- If `random_state` is `None` (or `np.random`), it would now be controlled by the global random state `np.random.seed` instead of being unpredictable.

Do you think that is OK?

Reply:
This seems to be getting too confusing IMO.

I intentionally wrote the tests `test_random_state` and `test_random_state_None_int` when this was coded years ago for the reasons noted in the tests there:

> ensure that the global random state is not modified because
> the directed Hausdorff algorithm uses randomization

and

> check that seed values of None or int do not alter global
> random state

I think the original design was meant to keep the function insulated from affecting/consuming external random states.

If I understand Robert correctly, it may make sense to deviate from the SPEC here and keep things mostly as they are with the `seed` argument and just transition that to `Generator`. That said, I'd still prefer to deprecate the switch.

Reply:
(or whatever class of warning is needed... I guess you can't really deprecate `seed=0` changing that easily since it can't be suppressed by user cooperation..

Reply:
I would recommend (ultimately) going through with the SPEC 7 conventions (the argument name being `rng`, using a `Generator` internally). I would skip the intermediate stage where we `check_random_state()` can use the legacy global state. If that means that you don't get to use the standard decorator in this one place, that's fine.

Reply:
I brought this up because the PR breaks backward compatibility immediately.
```python3
seed = np.int64(1)
directed_hausdorff(u, v, seed=seed)  # works before, but not with this PR, failing with
# AttributeError: 'numpy.int64' object has no attribute 'shuffle'
```
To avoid this without looking for all possible types that `seed` used to accept, the first paragraph of my comment is to change this to:
```suggestion
    if isinstance(seed, np.random.Generator):
    # Generator passthrough
        rng = seed
    else:
        rng = np.random.RandomState(seed)
```

---

> I think the original design was meant to keep the function insulated from affecting/consuming external random states.

If that is a very important property to preserve, that's fine. If it were *not* important to preserve that, I was suggesting that this be temporarily simplified (during the SPEC 7 deprecation period) to:
```suggestion
    rng = check_random_state(seed)
```
for the reason that it would make the SPEC 7 transition simpler.

But again, I don't know anything about this function or the existing tests, so if it's important to not affect/consume the external random state, that's fine. I just wanted to explore options to keep things easier for us.

Reply:
Yes, it's an important property to preserve.